File: services/papertrail.rb
class Papertrail < Grape::API
  namespace :papertrail do
    helpers do
      def event
        @event ||= Hashie::Mash.new JSON.parse(params.payload)
      end
    end

    post do
      title = "Search <#{event.saved_search.html_search_url}|#{event.saved_search.name}> has triggered!\n"

      payload = {
        username: 'Papertrail',
        text: title + event.events.map { |event| "#{event.hostname}: #{event.message}" }.join("\n")
      }

      post payload
    end
  end
end

File: services/papertrail/papertrail.go
package papertrail

import (
	"encoding/json"
	"github.com/remind101/slack-hooks/slack"
)

type Event struct {
	Hostname string `json:"hostname"`
	Message  string `json:"message"`
}

type Payload struct {
	Events []Event `json:"events"`
}

type Service struct {
}

func (s *Service) Parse(input string) interface{} {
	payload := &Payload{}

	if err := json.Unmarshal([]byte(input), payload); err != nil {
		panic(err)
	}

	return payload
}

func (s *Service) Handle(payload interface{}) *slack.Message {
	return &slack.Message{
		Username: "Papertrail",
		Text:     "Search",
	}
}
